Use `utils.data` to handle compressed ecsv

kYwzor opened the pull request:
### Description
This pull request is to change `io.misc.ecsv` to handle compressed data via `utils.data`, rather than having its own implementation. This expands its compression support to include LZMA- and LZW-compressed files while also keeping its support of gzip and bz2 files.

Fixes #18490.

<!-- Optional opt-out -->

- [ ] By checking this box, the PR author has requested that maintainers do **NOT** use the "Squash and Merge" button. Maintainers should respect this when possible; however, the final decision is at the discretion of the maintainer that merges the PR.

kYwzor commented:
Added a test for opening ECSV files compressed with gzip, bz2 and lzma, using all the available engines. If CI passes, I think this is ready for final review and merge.

kYwzor commented:
Interestingly, the test case I added was triggering a bug with the pyarrow engine, that was technically not introduced by my changes. Basically, our pyarrow code wasn't really considering the possibility of compressed files. I made a new commit changing that code to also use `get_readable_fileobj` and I think that will do that trick. Fingers crossed for green CI :)

kYwzor commented:
The associated test will definitely not work in a backport, as the old code did not support lzma files (.xz).

taldcroft commented on astropy/io/misc/ecsv.py:
```diff
@@ -463,25 +462,8 @@ def get_header_lines(
-            tmp_input_file = input_file
-
-        for idx, line in enumerate(tmp_input_file):
+    with get_readable_fileobj(input_file, encoding="binary") as f:
```
Nice!

taldcroft commented on astropy/io/misc/pyarrow/csv.py:
```diff
@@ -159,12 +159,13 @@ def read_csv(
-        read_options=read_options,
-        convert_options=convert_options,
-    )
+    with get_readable_fileobj(input_file, encoding="binary") as f:
```
Not knowing about the pyarrow internals, I'm wondering if this might have a performance impact for the case of a uncompressed file supplied as the file name (str or Path). As a sanity check, can you measure the read time for a large uncompressed CSV file with current master and your new code?

taldcroft reviewed (changes requested):
@kYwzor - this looks great, thanks! I have a few comments and minor change requests.

taldcroft commented:
@kYwzor - I had not noticed previously the failing test on Windows (initially thought it was not a required test). This would seem to be an upstream issue in pyarrow (Windows-only). Since this is not introduced by this PR, I would suggest to mark the failing tests as `xfail` for Windows and we can merge this PR.

After that it would be worth you moving https://github.com/astropy/astropy/pull/18712#issuecomment-3427237649 out to a new issue. There is probably a mix of resolutions there, including upstream problems or astropy (the fast ascii reader does not support unicode).

kYwzor commented:
Huh, I didn't notice it was only failing on Windows, that's interesting. I added the xfail, as you suggested, and squashed the commits. I'll create a separate ticket for the roundtrip issue, hopefully someone can fix that. If CI passes, it should be ok to merge.
